File: omrchecker-js/packages/core/src/processors/image/markerDetection.ts
/**
 * Marker Detection Module
 *
 * TypeScript port of src/processors/image/marker_detection.py
 *
 * Extracted from CropOnCustomMarkers to provide focused, testable
 * marker template matching algorithms.
 *
 * This module handles:
 * - Marker template preparation
 * - Multi-scale template matching
 * - Best match selection
 * - Marker corner extraction
 */

import cv from '../../utils/opencv';
import { Logger } from '../../utils/logger';
import { MathUtils } from '../../utils/math';

const logger = new Logger('MarkerDetection');

export interface ReferenceZone {
  origin: [number, number];
  dimensions: [number, number];
}

export interface MatchResult {
  position: [number, number] | null;
  optimalMarker: cv.Mat | null;
  confidence: number;
  scalePercent: number | null;
}

export interface DetectionResult {
  corners: number[][] | null;
}

/**
 * Extract and prepare marker template from reference image.
 *
 * Applies preprocessing to enhance marker features:
 * 1. Extract region of interest
 * 2. Resize if dimensions specified
 * 3. Gaussian blur to reduce noise
 * 4. Normalize to full range
 * 5. Optional erode-subtract to enhance edges
 *
 * @param referenceImage - Source image containing the marker
 * @param referenceZone - Object with 'origin' and 'dimensions' keys
 * @param markerDimensions - Optional [width, height] to resize marker
 * @param blurKernel - Gaussian blur kernel size
 * @param applyErodeSubtract - Whether to enhance edges with erosion
 * @returns Preprocessed marker template ready for matching
 */
export function prepareMarkerTemplate(
  referenceImage: cv.Mat,
  referenceZone: ReferenceZone,
  markerDimensions?: [number, number],
  blurKernel: [number, number] = [5, 5],
  applyErodeSubtract: boolean = true
): cv.Mat {
  const { origin, dimensions } = referenceZone;
  const [x, y] = origin;
  const [w, h] = dimensions;

  // Extract marker region
  const rect = new cv.Rect(x, y, w, h);
  let marker = referenceImage.roi(rect);

  try {
    // Resize if dimensions specified
    if (markerDimensions) {
      const [targetW, targetH] = markerDimensions;
      const resized = new cv.Mat();
      cv.resize(marker, resized, new cv.Size(targetW, targetH));
      marker.delete();
      marker = resized;
    }

    // Blur to reduce noise
    const blurred = new cv.Mat();
    const ksize = new cv.Size(blurKernel[0], blurKernel[1]);
    cv.GaussianBlur(marker, blurred, ksize, 0);
    marker.delete();
    marker = blurred;

    // Normalize to full range
    let normalized = new cv.Mat();
    cv.normalize(marker, normalized, 0, 255, cv.NORM_MINMAX, cv.CV_8U);
    marker.delete();
    marker = normalized;

    // Optional edge enhancement
    if (applyErodeSubtract) {
      // Erode
      const eroded = new cv.Mat();
      const kernel = cv.Mat.ones(5, 5, cv.CV_8U);
      cv.erode(marker, eroded, kernel, new cv.Point(-1, -1), 5);
      kernel.delete();

      // Subtract
      const subtracted = new cv.Mat();
      cv.subtract(marker, eroded, subtracted);
      eroded.delete();
      marker.delete();
      marker = subtracted;

      // Re-normalize after edge enhancement
      normalized = new cv.Mat();
      cv.normalize(marker, normalized, 0, 255, cv.NORM_MINMAX, cv.CV_8U);
      marker.delete();
      marker = normalized;
    }

    return marker;
  } catch (error) {
    marker.delete();
    throw error;
  }
}

/**
 * Perform multi-scale template matching to find best match.
 *
 * Tests marker at different scales within scaleRange to account for
 * size variations due to scanning/printing differences.
 *
 * @param patch - Image patch to search in
 * @param marker - Template marker to find
 * @param scaleRange - [minPercent, maxPercent] for scaling
 * @param scaleSteps - Number of scale increments to test
 * @returns Match result with position, marker, confidence, and scale
 */
export function multiScaleTemplateMatch(
  patch: cv.Mat,
  marker: cv.Mat,
  scaleRange: [number, number] = [85, 115],
  scaleSteps: number = 5
): MatchResult {
  const descentPerStep = (scaleRange[1] - scaleRange[0]) / scaleSteps;
  const markerHeight = marker.rows;
  const markerWidth = marker.cols;
  const patchHeight = patch.rows;
  const patchWidth = patch.cols;

  let bestPosition: [number, number] | null = null;
  let bestMarker: cv.Mat | null = null;
  let bestConfidence = 0.0;
  let bestScalePercent: number | null = null;

  // Test different scales
  for (
    let scalePercent = scaleRange[1];
    scalePercent > scaleRange[0];
    scalePercent -= descentPerStep
  ) {
    const scale = scalePercent / 100;
    if (scale <= 0.0) continue;

    // Rescale marker
    const scaledWidth = Math.round(markerWidth * scale);
    const scaledHeight = Math.round(markerHeight * scale);

    // Skip if rescaled marker is larger than patch
    if (scaledHeight > patchHeight || scaledWidth > patchWidth) {
      continue;
    }

    const scaledMarker = new cv.Mat();
    try {
      cv.resize(marker, scaledMarker, new cv.Size(scaledWidth, scaledHeight), 0, 0, cv.INTER_LINEAR);

      // Template matching
      const matchResult = new cv.Mat();
      cv.matchTemplate(patch, scaledMarker, matchResult, cv.TM_CCOEFF_NORMED);

      const result = cv.minMaxLoc(matchResult, new cv.Mat());
      const maxConfidence = result.maxVal;

      matchResult.delete();

      // Update best match if this is better
      if (maxConfidence > bestConfidence) {
        bestScalePercent = scalePercent;
        if (bestMarker) {
          bestMarker.delete();
        }
        bestMarker = scaledMarker.clone();
        bestConfidence = maxConfidence;
        bestPosition = [result.maxLoc.x, result.maxLoc.y];
      }

      scaledMarker.delete();
    } catch (error) {
      scaledMarker.delete();
      throw error;
    }
  }

  logger.debug(
    `Best match: scale=${bestScalePercent}%, confidence=${(bestConfidence * 100).toFixed(2)}%`
  );

  return {
    position: bestPosition,
    optimalMarker: bestMarker,
    confidence: bestConfidence,
    scalePercent: bestScalePercent,
  };
}

/**
 * Extract corner points of detected marker.
 *
 * @param position - [x, y] position of marker's top-left corner in patch
 * @param marker - The marker template (for dimensions)
 * @param zoneOffset - Offset to convert from patch coordinates to absolute
 * @returns 4x2 array of corner points in absolute coordinates
 */
export function extractMarkerCorners(
  position: [number, number],
  marker: cv.Mat,
  zoneOffset: [number, number] = [0, 0]
): number[][] {
  const h = marker.rows;
  const w = marker.cols;
  const [x, y] = position;

  // Get rectangle corners in patch coordinates
  const corners = MathUtils.getRectanglePoints(x, y, w, h);

  // Shift to absolute coordinates
  const absoluteCorners = MathUtils.shiftPointsFromOrigin(zoneOffset, corners);

  return absoluteCorners;
}

/**
 * Detect marker in a patch using multi-scale template matching.
 *
 * Main entry point for marker detection. Combines:
 * 1. Multi-scale template matching
 * 2. Confidence threshold check
 * 3. Corner extraction
 *
 * @param patch - Image patch to search in
 * @param marker - Template marker to find
 * @param zoneOffset - Offset for absolute coordinate conversion
 * @param scaleRange - [minPercent, maxPercent] for scaling
 * @param scaleSteps - Number of scale increments to test
 * @param minConfidence - Minimum confidence threshold (0.0 to 1.0)
 * @returns 4x2 array of corner points in absolute coordinates, or null if not found
 */
export function detectMarkerInPatch(
  patch: cv.Mat,
  marker: cv.Mat,
  zoneOffset: [number, number] = [0, 0],
  scaleRange: [number, number] = [85, 115],
  scaleSteps: number = 5,
  minConfidence: number = 0.3
): number[][] | null {
  // Perform multi-scale matching
  const matchResult = multiScaleTemplateMatch(patch, marker, scaleRange, scaleSteps);

  // Check if we found a valid match
  if (!matchResult.position || !matchResult.optimalMarker) {
    logger.warn('No marker match found in patch');
    if (matchResult.optimalMarker) {
      matchResult.optimalMarker.delete();
    }
    return null;
  }

  if (matchResult.confidence < minConfidence) {
    logger.warn(
      `Marker match confidence ${(matchResult.confidence * 100).toFixed(2)}% below threshold ${(
        minConfidence * 100
      ).toFixed(2)}%`
    );
    matchResult.optimalMarker.delete();
    return null;
  }

  // Extract corners
  const corners = extractMarkerCorners(
    matchResult.position,
    matchResult.optimalMarker,
    zoneOffset
  );

  matchResult.optimalMarker.delete();

  logger.debug(
    `Marker detected at [${matchResult.position}] with confidence ${(
      matchResult.confidence * 100
    ).toFixed(2)}% (scale=${matchResult.scalePercent}%)`
  );

  return corners;
}

/**
 * Validate detected marker corners.
 *
 * @param corners - 4x2 array of corner points
 * @param expectedAreaRange - Optional [minArea, maxArea] for validation
 * @returns True if corners are valid, false otherwise
 */
export function validateMarkerDetection(
  corners: number[][] | null,
  expectedAreaRange?: [number, number]
): boolean {
  if (!corners) {
    return false;
  }

  if (corners.length !== 4 || corners.some((c) => c.length !== 2)) {
    logger.warn(`Invalid corner shape: expected 4x2, got ${corners.length}`);
    return false;
  }

  // Optional area validation
  if (expectedAreaRange) {
    // Calculate area using cross product
    // Convert to cv.Mat for contourArea
    const mat = cv.matFromArray(corners.length, 1, cv.CV_32SC2, corners.flat());
    const area = cv.contourArea(mat);
    mat.delete();

    const [minArea, maxArea] = expectedAreaRange;

    if (area < minArea || area > maxArea) {
      logger.warn(`Marker area ${area} outside expected range [${minArea}, ${maxArea}]`);
      return false;
    }
  }

  return true;
}

File: src/processors/image/crop_on_patches/marker_detection.py
"""
Marker Detection Module

Extracted from CropOnCustomMarkers to provide focused, testable
marker template matching algorithms.

This module handles:
- Marker template preparation
- Multi-scale template matching
- Best match selection
- Marker corner extraction
"""

from typing import Optional, Tuple
import cv2
import numpy as np

from src.utils.image import ImageUtils
from src.utils.math import MathUtils
from src.utils.logger import logger


def prepare_marker_template(
    reference_image: np.ndarray,
    reference_zone: dict,
    marker_dimensions: Optional[Tuple[int, int]] = None,
    blur_kernel: Tuple[int, int] = (5, 5),
    apply_erode_subtract: bool = True,
) -> np.ndarray:
    """
    Extract and prepare marker template from reference image.

    Applies preprocessing to enhance marker features:
    1. Extract region of interest
    2. Resize if dimensions specified
    3. Gaussian blur to reduce noise
    4. Normalize to full range
    5. Optional erode-subtract to enhance edges

    Args:
        reference_image: Source image containing the marker
        reference_zone: Dict with 'origin' and 'dimensions' keys
        marker_dimensions: Optional (width, height) to resize marker
        blur_kernel: Gaussian blur kernel size
        apply_erode_subtract: Whether to enhance edges with erosion

    Returns:
        Preprocessed marker template ready for matching
    """
    origin, dimensions = reference_zone["origin"], reference_zone["dimensions"]
    x, y = origin
    w, h = dimensions

    # Extract marker region
    marker = reference_image[y : y + h, x : x + w]

    # Resize if dimensions specified
    if marker_dimensions is not None:
        marker = ImageUtils.resize_to_dimensions(marker_dimensions, marker)

    # Blur to reduce noise
    marker = cv2.GaussianBlur(marker, blur_kernel, 0)

    # Normalize to full range
    marker = cv2.normalize(
        marker, None, alpha=0, beta=255, norm_type=cv2.NORM_MINMAX, dtype=cv2.CV_8U
    )

    # Optional edge enhancement
    if apply_erode_subtract:
        marker = marker - cv2.erode(marker, kernel=np.ones((5, 5)), iterations=5)
        # Re-normalize after edge enhancement
        marker = cv2.normalize(
            marker, None, alpha=0, beta=255, norm_type=cv2.NORM_MINMAX, dtype=cv2.CV_8U
        )

    return marker


def multi_scale_template_match(
    patch: np.ndarray,
    marker: np.ndarray,
    scale_range: Tuple[int, int] = (85, 115),
    scale_steps: int = 5,
) -> Tuple[Optional[np.ndarray], Optional[np.ndarray], float, Optional[int]]:
    """
    Perform multi-scale template matching to find best match.

    Tests marker at different scales within scale_range to account for
    size variations due to scanning/printing differences.

    Args:
        patch: Image patch to search in
        marker: Template marker to find
        scale_range: (min_percent, max_percent) for scaling
        scale_steps: Number of scale increments to test

    Returns:
        Tuple of (position, optimal_marker, confidence, optimal_scale_percent)
        - position: (x, y) of top-left corner, or None if not found
        - optimal_marker: Rescaled marker that gave best match
        - confidence: Match confidence (0.0 to 1.0)
        - optimal_scale_percent: Scale that gave best match
    """
    descent_per_step = (scale_range[1] - scale_range[0]) // scale_steps
    marker_height, marker_width = marker.shape[:2]
    patch_height, patch_width = patch.shape[:2]

    best_position = None
    best_marker = None
    best_confidence = 0.0
    best_scale_percent = None

    # Test different scales
    for scale_percent in np.arange(
        scale_range[1], scale_range[0], -1 * descent_per_step
    ):
        scale = float(scale_percent / 100)
        if scale <= 0.0:
            continue

        # Rescale marker
        scaled_marker = ImageUtils.resize_single(
            marker,
            u_width=int(marker_width * scale),
            u_height=int(marker_height * scale),
        )

        # Skip if rescaled marker is larger than patch
        scaled_height, scaled_width = scaled_marker.shape[:2]
        if scaled_height > patch_height or scaled_width > patch_width:
            continue

        # Template matching
        match_result = cv2.matchTemplate(patch, scaled_marker, cv2.TM_CCOEFF_NORMED)

        max_confidence = match_result.max()

        # Update best match if this is better
        if max_confidence > best_confidence:
            best_scale_percent = scale_percent
            best_marker = scaled_marker
            best_confidence = max_confidence

            # Get position of best match
            min_val, max_val, min_loc, max_loc = cv2.minMaxLoc(match_result)
            best_position = max_loc  # (x, y)

    logger.debug(
        f"Best match: scale={best_scale_percent}%, confidence={best_confidence:.2%}"
    )

    return best_position, best_marker, best_confidence, best_scale_percent


def extract_marker_corners(
    position: Tuple[int, int],
    marker: np.ndarray,
    zone_offset: Tuple[int, int] = (0, 0),
) -> np.ndarray:
    """
    Extract corner points of detected marker.

    Args:
        position: (x, y) position of marker's top-left corner in patch
        marker: The marker template (for dimensions)
        zone_offset: Offset to convert from patch coordinates to absolute

    Returns:
        4x2 array of corner points in absolute coordinates
    """
    h, w = marker.shape[:2]
    x, y = position

    # Get rectangle corners in patch coordinates
    corners = MathUtils.get_rectangle_points(x, y, w, h)

    # Shift to absolute coordinates
    absolute_corners = MathUtils.shift_points_from_origin(zone_offset, corners)

    # Ensure we return a numpy array
    return np.array(absolute_corners, dtype=np.float32)


def detect_marker_in_patch(
    patch: np.ndarray,
    marker: np.ndarray,
    zone_offset: Tuple[int, int] = (0, 0),
    scale_range: Tuple[int, int] = (85, 115),
    scale_steps: int = 5,
    min_confidence: float = 0.3,
) -> Optional[np.ndarray]:
    """
    Detect marker in a patch using multi-scale template matching.

    Main entry point for marker detection. Combines:
    1. Multi-scale template matching
    2. Confidence threshold check
    3. Corner extraction

    Args:
        patch: Image patch to search in
        marker: Template marker to find
        zone_offset: Offset for absolute coordinate conversion
        scale_range: (min_percent, max_percent) for scaling
        scale_steps: Number of scale increments to test
        min_confidence: Minimum confidence threshold (0.0 to 1.0)

    Returns:
        4x2 array of corner points in absolute coordinates, or None if not found
    """
    # Perform multi-scale matching
    position, optimal_marker, confidence, scale_percent = multi_scale_template_match(
        patch, marker, scale_range, scale_steps
    )

    # Check if we found a valid match
    if position is None or optimal_marker is None:
        logger.warning("No marker match found in patch")
        return None

    if confidence < min_confidence:
        logger.warning(
            f"Marker match confidence {confidence:.2%} below threshold {min_confidence:.2%}"
        )
        return None

    # Extract corners
    corners = extract_marker_corners(position, optimal_marker, zone_offset)

    logger.debug(
        f"Marker detected at {position} with confidence {confidence:.2%} "
        f"(scale={scale_percent}%)"
    )

    return corners


def validate_marker_detection(
    corners: Optional[np.ndarray],
    expected_area_range: Optional[Tuple[float, float]] = None,
) -> bool:
    """
    Validate detected marker corners.

    Args:
        corners: 4x2 array of corner points
        expected_area_range: Optional (min_area, max_area) for validation

    Returns:
        True if corners are valid, False otherwise
    """
    if corners is None:
        return False

    if corners.shape != (4, 2):
        logger.warning(f"Invalid corner shape: {corners.shape}, expected (4, 2)")
        return False

    # Optional area validation
    if expected_area_range is not None:
        # Calculate area using cross product
        # Assumes corners are ordered
        area = cv2.contourArea(corners)
        min_area, max_area = expected_area_range

        if not (min_area <= area <= max_area):
            logger.warning(
                f"Marker area {area} outside expected range [{min_area}, {max_area}]"
            )
            return False

    return True
